## while: warming up
The while loop is like a repeated if statement. The code is executed over and over again, as long as the condition is True. Have another look at its recipe.

***while condition :*** <br>
    &emsp;***expression*** <br> 

Can you tell how many printouts the following while loop will do?

***x = 1<br>
while x < 4 :<br>
    &emsp;print(x)<br>
   &emsp;x = x + 1***<br>


### Instructions

Possible Answers

- [ ] 0

- [ ] 1

- [ ] 2

- [x] 3

- [ ] 4

## Basic while loop
Below you can find the example from the video where the **error** variable, initially equal to **50.0**, is divided by 4 and printed out on every run:

***error = 50.0 <br>
while error > 1 : <br>
    &emsp;error = error / 4 <br>
    &emsp;print(error)*** <br>

This example will come in handy, because it's time to build a **while** loop yourself! We're going to code a **while** loop that implements a very basic control system for an **inverted pendulum**. If there's an offset from standing perfectly straight, the **while** loop will incrementally fix this offset.

Note that if your **while** loop takes too long to run, you might have made a mistake. In particular, remember to **indent** the contents of the loop!

### Instructions

- Create the variable **offset** with an initial value of **8**.
- Code a **while** loop that keeps running as long as **offset** is not equal to **0**. Inside the **while** loop:
  - Print out the sentence **"correcting..."**.
  - Next, decrease the value of **offset** by 1. You can do this with **offset = offset - 1**.
  - Finally, still within your loop, print out **offset** so you can see how it changes.

In [1]:
offset = 8

while offset !=0:
    print("correcting...")
    offset -= 1
    print(offset)

correcting...
7
correcting...
6
correcting...
5
correcting...
4
correcting...
3
correcting...
2
correcting...
1
correcting...
0


## Add conditionals
The **while** loop that corrects the **offset** is a good start, but what if **offset** is negative? You can try to run the following code where **offset** is initialized to **-6**:

*Initialize offset* <br>
***offset = -6***<br>

*Code the while loop*<br>
***while offset != 0 :<br>
    &emsp;print("correcting...")<br>
    &emsp;offset = offset - 1<br>
    &emsp;print(offset)***<br>

but your session will be disconnected. The **while** loop will never stop running, because **offset** will be further decreased on every run. **offset != 0** will never become **False** and the **while** loop continues forever.

Fix things by putting an **if-else** statement inside the **while** loop. If your code is still taking too long to run, you probably made a mistake!

### Instructions

- Inside the **while** loop, complete the **if-else** statement:
  - If **offset** is greater than zero, you should decrease **offset** by 1.
  - Else, you should increase **offset** by 1.
- If you've coded things correctly, hitting Submit Answer should work this time.
*If your code is still taking too long to run (or your session is expiring), you probably made a mistake. Check your code and make sure that the statement **offset != 0** will eventually evaluate to **FALSE**!*

In [2]:
offset = -6

while offset !=0:
    print("correcting...")
    if offset > 1:
        offset -= 1
    else:
        offset += 1
    print(offset)

correcting...
-5
correcting...
-4
correcting...
-3
correcting...
-2
correcting...
-1
correcting...
0


## Loop over a list
Have another look at the **for** loop that Hugo showed in the video:

***fam = [1.73, 1.68, 1.71, 1.89] <br>
for height in fam : <br>
    &emsp;print(height)***<br>

As usual, you simply have to indent the code with 4 spaces to tell Python which code should be executed in the **for** loop.

The **areas** variable, containing the area of different rooms in your house, is already defined.

### Instructions

- Write a **for** loop that iterates over all elements of the **areas** list and prints out every element separately.

In [3]:
# areas list
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Code the for loop
for i in areas :
    print(i)

11.25
18.0
20.0
10.75
9.5


## Indexes and values (1)
Using a **for** loop to iterate over a list only gives you access to every list element in each run, one after the other. If you also want to access the index information, so where the list element you're iterating over is located, you can use **enumerate()**.

As an example, have a look at how the **for** loop from the video was converted:

***fam = [1.73, 1.68, 1.71, 1.89] <br>
for index, height in enumerate(fam) :<br>
    &emsp;print("person " + str(index) + ": " + str(height))***<br>

### Instructions

- Adapt the **for** loop in the sample code to use **enumerate()** and use two iterator variables.
- Update the **print()** statement so that on each run, a line of the form **"room x: y"** should be printed, where x is the index of the list element and y is the actual list element, i.e. the area. Make sure to print out this exact string, with the correct spacing.

In [6]:
# areas list
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Change for loop to use enumerate() and update print()
for i, a in enumerate(areas) :
    print("room ",str(i),": ",str(a))

room  0 :  11.25
room  1 :  18.0
room  2 :  20.0
room  3 :  10.75
room  4 :  9.5


## Indexes and values (2)
For non-programmer folks, **room 0: 11.25** is strange. Wouldn't it be better if the count started at 1?

### Instructions

Adapt the **print()** function in the for loop so that the first printout becomes **"room 1: 11.25"**, the second one **"room 2: 18.0"** and so on.

In [7]:
# areas list
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Change for loop to use enumerate() and update print()
for i, a in enumerate(areas) :
    print("room ",str(i+1),": ",str(a))

room  1 :  11.25
room  2 :  18.0
room  3 :  20.0
room  4 :  10.75
room  5 :  9.5


## Loop over list of lists
Remember the **house** variable from the Intro to Python course? Have a look at its definition in the script. It's basically a list of lists, where each sublist contains the name and area of a room in your house.

It's up to you to build a **for** loop from scratch this time!

### Instructions

- Write a **for** loop that goes through each sublist of **house** and prints out **the x is y sqm**, where x is the name of the room and y is the area of the room.

In [10]:
# house list of lists
house = [["hallway", 11.25], 
         ["kitchen", 18.0], 
         ["living room", 20.0], 
         ["bedroom", 10.75], 
         ["bathroom", 9.50]]
         
# Build a for loop from scratch
for i in house:
    print('the '+str(i[0])+' is '+str(i[1])+' sqm')

the hallway is 11.25 sqm
the kitchen is 18.0 sqm
the living room is 20.0 sqm
the bedroom is 10.75 sqm
the bathroom is 9.5 sqm


## Loop over dictionary
In Python 3, you need the **items()** method to loop over a dictionary:

***world = { "afghanistan":30.55,<br>
         &emsp; "albania":2.77,<br>
         &emsp; "algeria":39.21 }***<br>

***for key, value in world.items() :<br>
    &emsp;print(key + " -- " + str(value))***<br>
    
Remember the **europe** dictionary that contained the names of some European countries as key and their capitals as corresponding value? Go ahead and write a loop to iterate over it!

### Instructions

Write a **for** loop that goes through each key:value pair of **europe**. On each iteration, **"the capital of x is y"** should be printed out, where x is the key and y is the value of the pair.

In [1]:
# Definition of dictionary
europe = {'spain':'madrid', 'france':'paris', 'germany':'berlin',
          'norway':'oslo', 'italy':'rome', 'poland':'warsaw', 'austria':'vienna' }
          
# Iterate over europe
for k,v in europe.items():
    print("the capital of",str(k),"is",str(v))

the capital of spain is madrid
the capital of france is paris
the capital of germany is berlin
the capital of norway is oslo
the capital of italy is rome
the capital of poland is warsaw
the capital of austria is vienna


## Loop over NumPy array
If you're dealing with a 1D NumPy array, looping over all elements can be as simple as:

***for x in my_array :<br>
    &emsp;...***<br>

If you're dealing with a 2D NumPy array, it's more complicated. A 2D array is built up of multiple 1D arrays. To explicitly iterate over all separate elements of a multi-dimensional array, you'll need this syntax:

***for x in np.nditer(my_array) :<br>
    &emsp;...***<br>

Two NumPy arrays that you might recognize from the intro course are available in your Python session: **np_height**, a NumPy array containing the heights of Major League Baseball players, and **np_baseball**, a 2D NumPy array that contains both the heights (first column) and weights (second column) of those players.

### Instructions

- Import the **numpy** package under the local alias **np**.
- Write a **for** loop that iterates over all elements in **np_height** and prints out **"x inches"** for each element, where x is the value in the array.
- Write a **for** loop that visits every element of the **np_baseball** array and prints it out.

In [ ]:
import numpy as np

np_height = np.array([74,74,	72,	72,	73,	69,	69,	71,	76,	71,	73,	73,	74,	74,	69,	70,	73,	75,	78,	79,	76,	74,	76,	72,	71,	75,	77,	74,	73,	74,	78,	73,	75,	73,	75,	75,	74,	69,	71,	74,	73,	73,	76,	74,	74,	70,	72,	77,	74,	70,	73,	75,	76,	76,	78,	74,	74,	76,	77,	81,	78,	75,	77,	75,	76,	74,	72,	72,	75,	73,	73,	73,	70,	70,	70,	76,	68,	71,	72,	75,	75,	75,	75,	68,	74,	78,	71,	73,	76,	74,	74,	79,	75,	73,	76,	74,	74,	73,	72,	74,	73,	74,	72,	73,	69,	72,	73,	75,	75,	73,	72,	72,	76,	74,	72,	77,	74,	77,	75,	76,	80,	74,	74,	75,	78,	73,	73,	74,	75,	76,	71,	73,	74,	76,	76,	74,	73,	74,	70,	72,	73,	73,	73,	73,	71,	74,	74,	72,	74,	71,	74,	73,	75,	75,	79,	73,	75,	76,	74,	76,	78,	74,	76,	72,	74,	76,	74,	75,	78,	75,	72,	74,	72,	74,	70,	71,	70,	75,	71,	71,	73,	72,	71,	73,	72,	75,	74,	74,	75,	73,	77,	73,	76,	75,	74,	76,	75,	73,	71,	76,	75,	72,	71,	77,	73,	74,	71,	72,	74,	75,	73,	72,	75,	75,	74,	72,	74,	71,	70,	74,	77,	77,	75,	75,	78,	75,	76,	73,	75,	75,	79,	77,	76,	71,	75,	74,	69,	71,	76,	72,	72,	70,	72,	73,	71,	72,	71,	73,	72,	73,	74,	74,	72,	75,	74,	74,	77,	75,	73,	72,	71,	74,	77,	75,	75,	75,	78,	78,	74,	76,	78,	76,	70,	72,	80,	74,	74,	71,	70,	72,	71,	74,	71,	72,	71,	74,	69,	76,	75,	75,	76,	73,	76,	73,	77,	73,	72,	72,	77,	77,	71,	74,	74,	73,	78,	75,	73,	70,	74,	72,	73,	73,	75,	75,	74,	76,	73,	74,	75,	75,	72,	73,	73,	72,	74,	78,	76,	73,	74,	75,	70,	75,	71,	72,	78,	75,	73,	73,	71,	75,	77,	72,	69,	73,	74,	72,	70,	75,	70,	72,	72,	74,	73,	74,	76,	75,	80,	72,	75,	73,	74,	74,	73,	75,	75,	71,	73,	75,	74,	74,	72,	74,	74,	74,	73,	76,	75,	72,	73,	73,	73,	72,	72,	72,	72,	71,	75,	75,	74,	73,	75,	79,	74,	76,	73,	74,	74,	72,	74,	74,	75,	78,	74,	74,	74,	77,	70,	73,	74,	73,	71,	75,	71,	72,	77,	74,	70,	77,	73,	72,	76,	71,	76,	78,	75,	73,	78,	74,	79,	75,	76,	72,	75,	75,	70,	72,	70,	74,	71,	76,	73,	76,	71,	69,	72,	72,	69,	73,	69,	73,	74,	74,	72,	71,	72,	72,	76,	76,	76,	74,	76,	75,	71,	72,	71,	73,	75,	76,	75,	71,	75,	74,	72,	73,	73,	73,	73,	76,	72,	76,	73,	73,	73,	75,	75,	77,	73,	72,	75,	70,	74,	72,	80,	71,	71,	74,	74,	73,	75,	76,	73,	77,	72,	73,	77,	76,	71,	75,	73,	74,	77,	71,	72,	73,	69,	73,	70,	74,	76,	73,	73,	75,	73,	79,	74,	73,	74,	77,	75,	74,	73,	77,	73,	77,	74,	74,	73,	77,	74,	77,	75,	77,	75,	71,	74,	70,	79,	72,	72,	70,	74,	74,	72,	73,	72,	74,	74,	76,	82,	74,	74,	70,	73,	73,	74,	77,	72,	76,	73,	73,	72,	74,	74,	71,	72,	75,	74,	74,	77,	70,	71,	73,	76,	71,	75,	74,	72,	76,	79,	76,	73,	76,	78,	75,	76,	72,	72,	73,	73,	75,	71,	76,	70,	75,	74,	75,	73,	71,	71,	72,	73,	73,	72,	69,	73,	78,	71,	73,	75,	76,	70,	74,	77,	75,	79,	72,	77,	73,	75,	75,	75,	73,	73,	76,	77,	75,	70,	71,	71,	75,	74,	69,	70,	75,	72,	75,	73,	72,	72,	72,	76,	75,	74,	69,	73,	72,	72,	75,	77,	76,	80,	77,	76,	79,	71,	75,	73,	76,	77,	73,	76,	70,	75,	73,	75,	70,	69,	71,	72,	72,	73,	70,	70,	73,	76,	75,	72,	73,	79,	71,	72,	74,	74,	74,	72,	76,	76,	72,	72,	71,	72,	72,	70,	77,	74,	72,	76,	71,	76,	71,	73,	70,	73,	73,	72,	71,	71,	71,	72,	72,	74,	74,	74,	71,	72,	75,	72,	71,	72,	72,	72,	72,	74,	74,	77,	75,	73,	75,	73,	76,	72,	77,	75,	72,	71,	71,	75,	72,	73,	73,	71,	70,	75,	71,	76,	73,	68,	71,	72,	74,	77,	72,	76,	78,	81,	72,	73,	76,	72,	72,	74,	76,	73,	76,	75,	70,	71,	74,	72,	73,	76,	76,	73,	71,	68,	71,	71,	74,	77,	69,	72,	76,	75,	76,	75,	76,	72,	74,	76,	74,	72,	75,	78,	77,	70,	72,	79,	74,	71,	68,	77,	75,	71,	72,	70,	72,	72,	73,	72,	74,	72,	72,	75,	72,	73,	74,	72,	78,	75,	72,	74,	75,	75,	76,	74,	74,	73,	74,	71,	74,	75,	76,	74,	76,	76,	73,	75,	75,	74,	68,	72,	75,	71,	70,	72,	73,	72,	75,	74,	70,	76,	71,	82,	72,	73,	74,	71,	75,	77,	72,	74,	72,	73,	78,	77,	73,	73,	73,	73,	73,	76,	75,	70,	73,	72,	73,	75,	74,	73,	73,	76,	73,	75,	70,	77,	72,	77,	74,	75,	75,	75,	75,	72,	74,	71,	76,	71,	75,	76,	83,	75,	74,	76,	72,	72,	75,	75,	72,	77,	73,	72,	70,	74,	72,	74,	72,	71,	70,	71,	76,	74,	76,	74,	74,	74,	75,	75,	71,	71,	74,	77,	71,	74,	75,	77,	76,	74,	76,	72,	71,	72,	75,	73,	68,	72,	69,	73,	73,	75,	70,	70,	74,	75,	74,	74,	73,	74,	75,	77,	73,	74,	76,	74,	75,	73,	76,	78,	75,	73,	77,	74,	72,	74,	72,	71,	73,	75,	73,	67,	67,	76,	74,	73,	70,	75,	70,	72,	77,	79,	78,	74,	75,	75,	78,	76,	75,	69,	75,	72,	75,	73,	74,	75,	75,	73])

np_baseball = np.empty((1015,2))

np_baseball[:,0] = np_height

np_baseball[:,1] = [180,	215,	210,	210,	188,	176,	209,	200,	231,	180,	188,	180,	185,	160,	180,	185,	189,	185,	219,	230,	205,	230,	195,	180,	192,	225,	203,	195,	182,	188,	200,	180,	200,	200,	245,	240,	215,	185,	175,	199,	200,	215,	200,	205,	206,	186,	188,	220,	210,	195,	200,	200,	212,	224,	210,	205,	220,	195,	200,	260,	228,	270,	200,	210,	190,	220,	180,	205,	210,	220,	211,	200,	180,	190,	170,	230,	155,	185,	185,	200,	225,	225,	220,	160,	205,	235,	250,	210,	190,	160,	200,	205,	222,	195,	205,	220,	220,	170,	185,	195,	220,	230,	180,	220,	180,	180,	170,	210,	215,	200,	213,	180,	192,	235,	185,	235,	210,	222,	210,	230,	220,	180,	190,	200,	210,	194,	180,	190,	240,	200,	198,	200,	195,	210,	220,	190,	210,	225,	180,	185,	170,	185,	185,	180,	178,	175,	200,	204,	211,	190,	210,	190,	190,	185,	290,	175,	185,	200,	220,	170,	220,	190,	220,	205,	200,	250,	225,	215,	210,	215,	195,	200,	194,	220,	180,	180,	170,	195,	180,	170,	206,	205,	200,	225,	201,	225,	233,	180,	225,	180,	220,	180,	237,	215,	190,	235,	190,	180,	165,	195,	200,	190,	190,	185,	185,	205,	190,	205,	206,	220,	208,	170,	195,	210,	190,	211,	230,	170,	185,	185,	241,	225,	210,	175,	230,	200,	215,	198,	226,	278,	215,	230,	240,	184,	219,	170,	218,	190,	225,	220,	176,	190,	197,	204,	167,	180,	195,	220,	215,	185,	190,	205,	205,	200,	210,	215,	200,	205,	211,	190,	208,	200,	210,	232,	230,	210,	220,	210,	202,	212,	225,	170,	190,	200,	237,	220,	170,	193,	190,	150,	220,	200,	190,	185,	185,	200,	172,	220,	225,	190,	195,	219,	190,	197,	200,	195,	210,	177,	220,	235,	180,	195,	195,	190,	230,	190,	200,	190,	190,	200,	200,	184,	200,	180,	219,	187,	200,	220,	205,	190,	170,	160,	215,	175,	205,	200,	214,	200,	190,	180,	205,	220,	190,	215,	235,	191,	200,	181,	200,	210,	240,	185,	165,	190,	185,	175,	155,	210,	170,	175,	220,	210,	205,	200,	205,	195,	240,	150,	200,	215,	202,	200,	190,	205,	190,	160,	215,	185,	200,	190,	210,	185,	220,	190,	202,	205,	220,	175,	160,	190,	200,	229,	206,	220,	180,	195,	175,	188,	230,	190,	200,	190,	219,	235,	180,	180,	180,	200,	234,	185,	220,	223,	200,	210,	200,	210,	190,	177,	227,	180,	195,	199,	175,	185,	240,	210,	180,	194,	225,	180,	205,	193,	230,	230,	220,	200,	249,	190,	208,	245,	250,	160,	192,	220,	170,	197,	155,	190,	200,	220,	210,	228,	190,	160,	184,	180,	180,	200,	176,	160,	222,	211,	195,	200,	175,	206,	240,	185,	260,	185,	221,	205,	200,	170,	201,	205,	185,	205,	245,	220,	210,	220,	185,	175,	170,	180,	200,	210,	175,	220,	206,	180,	210,	195,	200,	200,	164,	180,	220,	195,	205,	170,	240,	210,	195,	200,	205,	192,	190,	170,	240,	200,	205,	175,	250,	220,	224,	210,	195,	180,	245,	175,	180,	215,	175,	180,	195,	230,	230,	205,	215,	195,	180,	205,	180,	190,	180,	190,	190,	220,	210,	255,	190,	230,	200,	205,	210,	225,	215,	220,	205,	200,	220,	197,	225,	187,	245,	185,	185,	175,	200,	180,	188,	225,	200,	210,	245,	213,	231,	165,	228,	210,	250,	191,	190,	200,	215,	254,	232,	180,	215,	220,	180,	200,	170,	195,	210,	200,	220,	165,	180,	200,	200,	170,	224,	220,	180,	198,	240,	239,	185,	210,	220,	200,	195,	220,	230,	170,	220,	230,	165,	205,	192,	210,	205,	200,	210,	185,	195,	202,	205,	195,	180,	200,	185,	240,	185,	220,	205,	205,	180,	201,	190,	208,	240,	180,	230,	195,	215,	190,	195,	215,	215,	220,	220,	230,	195,	190,	195,	209,	204,	170,	185,	205,	175,	210,	190,	180,	180,	160,	235,	200,	210,	180,	190,	197,	203,	205,	170,	200,	250,	200,	220,	200,	190,	170,	190,	220,	215,	206,	215,	185,	235,	188,	230,	195,	168,	190,	160,	200,	200,	189,	180,	190,	200,	220,	187,	240,	190,	180,	185,	210,	220,	219,	190,	193,	175,	180,	215,	210,	200,	190,	185,	220,	170,	195,	205,	195,	210,	190,	190,	180,	220,	190,	186,	185,	190,	180,	190,	170,	210,	240,	220,	180,	210,	210,	195,	160,	180,	205,	200,	185,	245,	190,	210,	200,	200,	222,	215,	240,	170,	220,	156,	190,	202,	221,	200,	190,	210,	190,	200,	165,	190,	185,	230,	208,	209,	175,	180,	200,	205,	200,	250,	210,	230,	244,	202,	240,	200,	215,	177,	210,	170,	215,	217,	198,	200,	220,	170,	200,	230,	231,	183,	192,	167,	190,	180,	180,	215,	160,	205,	223,	175,	170,	190,	240,	175,	230,	223,	196,	167,	195,	190,	250,	190,	190,	190,	170,	160,	150,	225,	220,	209,	210,	176,	260,	195,	190,	184,	180,	195,	195,	219,	225,	212,	202,	185,	200,	209,	200,	195,	228,	210,	190,	212,	190,	218,	220,	190,	235,	210,	200,	188,	210,	235,	188,	215,	216,	220,	180,	185,	200,	210,	220,	185,	231,	210,	195,	200,	205,	200,	190,	250,	185,	180,	170,	180,	208,	235,	215,	244,	220,	185,	230,	190,	200,	180,	190,	196,	180,	230,	224,	160,	178,	205,	185,	210,	180,	190,	200,	257,	190,	220,	165,	205,	200,	208,	185,	215,	170,	235,	210,	170,	180,	170,	190,	150,	230,	203,	260,	246,	186,	210,	198,	210,	215,	180,	200,	245,	200,	192,	192,	200,	192,	205,	190,	186,	170,	197,	219,	200,	220,	207,	225,	207,	212,	225,	170,	190,	210,	230,	210,	200,	238,	234,	222,	200,	190,	170,	220,	223,	210,	215,	196,	175,	175,	189,	205,	210,	180,	180,	197,	220,	228,	190,	204,	165,	216,	220,	208,	210,	215,	195,	200,	215,	229,	240,	207,	205,	208,	185,	190,	170,	208,	225,	190,	225,	185,	180,	165,	240,	220,	212,	163,	215,	175,	205,	210,	205,	208,	215,	180,	200,	230,	211,	230,	190,	220,	180,	205,	190,	180,	205,	190,	195]


for i in np_height:
    print(str(i)+" inches")
    
for j in np.nditer(np_baseball):
    print(j,end="_")

## Loop over DataFrame (1)
Iterating over a Pandas DataFrame is typically done with the **iterrows()** method. Used in a for loop, every observation is iterated over and on every iteration the row label and actual row contents are available:

***for lab, row in brics.iterrows() : <br>
    &emsp;...***

In this and the following exercises you will be working on the **cars** DataFrame. It contains information on the cars per capita and whether people drive right or left for seven countries in the world.

### Instructions

- Write a **for** loop that iterates over the rows of **cars** and on each iteration perform two **print()** calls: one to print out the row label and one to print out all of the rows contents.

In [12]:
import pandas as pd
cars = pd.read_csv('../Dictionaries & Pandas/cars.csv',index_col=0)

for r,c in cars.iterrows():
    print(r)
    print(c)

US
cars_per_cap              809
country         United States
drives_right             True
Name: US, dtype: object
AUS
cars_per_cap          731
country         Australia
drives_right        False
Name: AUS, dtype: object
JAP
cars_per_cap      588
country         Japan
drives_right    False
Name: JAP, dtype: object
IN
cars_per_cap       18
country         India
drives_right    False
Name: IN, dtype: object
RU
cars_per_cap       200
country         Russia
drives_right      True
Name: RU, dtype: object
MOR
cars_per_cap         70
country         Morocco
drives_right       True
Name: MOR, dtype: object
EG
cars_per_cap       45
country         Egypt
drives_right     True
Name: EG, dtype: object


## Loop over DataFrame (2)
The row data that's generated by **iterrows()** on every run is a Pandas Series. This format is not very convenient to print out. Luckily, you can easily select variables from the Pandas Series using square brackets:

***for lab, row in brics.iterrows() :<br>
    &emsp;print(row['country'])***<br>

### Instructions

- Using the iterators **lab** and **row**, adapt the code in the for loop such that the first iteration prints out **"US: 809"**, the second iteration **"AUS: 731"**, and so on.
- The output should be in the form **"country: cars_per_cap"**. Make sure to print out this exact string (with the correct spacing).
- *You can use **str()** to convert your integer data to a string so that you can print it in conjunction with the country label.*

In [16]:
for lab, row in cars.iterrows():
    print(lab+": "+str(row["cars_per_cap"]))

US: 809
AUS: 731
JAP: 588
IN: 18
RU: 200
MOR: 70
EG: 45


## Add column (1)
In the video, Hugo showed you how to add the length of the country names of the **brics** DataFrame in a new column:

***for lab, row in brics.iterrows() :<br>
    &emsp;brics.loc[lab, "name_length"] = len(row["country"])***<br>
    
You can do similar things on the **cars** DataFrame.

### Instructions

- Use a **for** loop to add a new column, named **COUNTRY**, that contains a uppercase version of the country names in the **"country"** column. You can use the string method **upper()** for this.
- To see if your code worked, print out **cars**. Don't indent this code, so that it's not part of the **for** loop.

In [26]:
for x, y in cars.iterrows():
    cars.loc[x,"COUNTRY"] = y["country"].upper()

print(cars)

     cars_per_cap        country  drives_right        COUNTRY
US            809  United States          True  UNITED STATES
AUS           731      Australia         False      AUSTRALIA
JAP           588          Japan         False          JAPAN
IN             18          India         False          INDIA
RU            200         Russia          True         RUSSIA
MOR            70        Morocco          True        MOROCCO
EG             45          Egypt          True          EGYPT


## Add column (2)
Using **iterrows()** to iterate over every observation of a Pandas DataFrame is easy to understand, but not very efficient. On every iteration, you're creating a new Pandas Series.

If you want to add a column to a DataFrame by calling a function on another column, the **iterrows()** method in combination with a for loop is not the preferred way to go. Instead, you'll want to use apply().

Compare the **iterrows()** version with the **apply()** version to get the same result in the **brics** DataFrame:

***for lab, row in brics.iterrows() :<br>
    &emsp;brics.loc[lab, "name_length"] = len(row["country"])***<br>

***brics["name_length"] = brics["country"].apply(len)***

We can do a similar thing to call the upper() method on every name in the country column. However, upper() is a method, so we'll need a slightly different approach:

### Instructions

- Replace the **for** loop with a one-liner that uses **.apply(str.upper)**. The call should give the same result: a column **COUNTRY** should be added to **cars**, containing an uppercase version of the country names.
- As usual, print out **cars** to see the fruits of your hard labor

In [27]:
import pandas as pd
cars = pd.read_csv('../Dictionaries & Pandas/cars.csv',index_col=0)

cars["COUNTRY"] = cars['country'].apply(str.upper)

print(cars)

     cars_per_cap        country  drives_right        COUNTRY
US            809  United States          True  UNITED STATES
AUS           731      Australia         False      AUSTRALIA
JAP           588          Japan         False          JAPAN
IN             18          India         False          INDIA
RU            200         Russia          True         RUSSIA
MOR            70        Morocco          True        MOROCCO
EG             45          Egypt          True          EGYPT
